### PR TITLE
[ISSUE #9906] cache token to improve performance

### LINKED ISSUE #9906: [Feature Request] Cache token to improve performance 缓存Token来提高接口的性能
ISSUE #9859 从token生成的角度来优化服务端的性能表现，取得了一定的效果，但是token的验证依然是非常耗时的。

在实际使用的场景中，默认 auth.enable = true 开启鉴权，客户端携带token访问服务端的查询接口（无论是配置发现还是服务发现），在服务端 4C8G的配置下，实测QPS都在两千上下。如果关闭鉴权，QPS在一万三左右。

鉴权这一块最主要的耗时是要验证token的合法性，需要解析token。
现在有一个简单的设想，如果把token缓存起来，直接内存字符串比对而不是解析，是不是就能大幅提升性能了？
实测，直接内存字符串比对，同等配置QPS接近关闭鉴权的效果。
并且，缓存token的方案，我们在生产上已经稳定运行了超过一年的时间。
跟@hujun-w-2 大佬对方案进行过简单的讨论和review，现在考虑发起社区讨论是否将这个能力提交给社区。

实现方案简述：

- 服务端缓存token；
- 服务端新增一个实体类TokenEntity，主要字段就是token string 和 token的过期时间；
- 服务端维护两个数据接口，一个 userMap<username, TokenEntity>，一个tokenMap<tokenStr, TokenEntity>;
- 当用户访问login接口时，若userMap里面存在该user，返回已经存在的token；否则，创建一个token返回，并将该token对应的TokenEntity存入userMap和tokenMap。（将token与username绑定，目的是大幅度减少新的token生成）；
- 当用户携带token访问服务端的接口时，判断tokenMap中是否存在该token，若存在，则直接返回；如果不存在，再转入原来的token解析流程，解析出来的token合法，就存入tokenMap（不要存userMap）再返回，不合法就返回AccessException；
- 服务端是以集群模式部署的，在token有效期内，一个服务端进程对同一个username颁发的token是一样的，userMap的最大元素数目等于在服务端login的账号数目，tokenMap的最大元素数目等于账号数目乘以服务端节点数；
- 服务端对定时扫描userMap和tokenMap，清除已经过期的token记录；

代码实现简述：
1、新增CachedJwtTokenManager 和 DelegateTokenManager
2、配置文件新增 nacos.core.auth.plugin.nacos.token.cache.enable字段开关，默认关。
3、原生NacosAuthManager中的tokenManager 的bean改成DelegateTokenManager
4、在DelegateTokenManager中进行cache开关判断。如果不开启，走原始的JwtTokenManager逻辑。如果开启，走CachedJwtTokenManager 逻辑。

方案缺陷：
缓存Token一定会带来额外的内存开销

请大家看一下这个方案是否有实现的必要，或者需要改进的地方，谢谢。

### PR DESCRIPTION
**Pull request by MajorHe1**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

fix #9906 

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Review by CherishCai — DISMISSED** [review]
LGTM

**Review by hujun-w-2 — COMMENTED** [review]
1.参数是不是应该放到application.properties文件中
2.补充更多的单测，特别是CachedJwtTokenManager.java

**Review by hujun-w-2 — APPROVED** [review]

File: plugin-default-impl/src/test/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    @Mock
+    private Authentication authentication;
+    
+    @Mock
+    private NacosUser user;
+    
+    @Before
+    public void setUp() throws Exception {
+        cachedJwtTokenManager = new CachedJwtTokenManager();
+        injectObject("jwtTokenManager", jwtTokenManager);
```
Comment (CherishCai) [inline plugin-default-impl/src/test/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManagerTest.java:56]: 用构造器注入是不是就不用反射设置属性啦～

File: plugin-default-impl/src/main/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManager.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+ * @author majorhe
+ */
+@Component
+public class CachedJwtTokenManager implements TokenManager {
+    
+    //key: token string, value: token entity
+    private volatile Map<String, TokenEntity> tokenMap = new ConcurrentHashMap<>(1024);
+    
+    //key: username, value: token entity. cache token created by self.
+    private volatile Map<String, TokenEntity> userMap = new ConcurrentHashMap<>(128);
```
Comment (YunWZ) [inline plugin-default-impl/src/main/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManager.java:49]: `username` and `TokenEntity` is a one-to-one relationship?
I think the same user may have multiple tokens at the same time.

File: plugin-default-impl/src/main/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManager.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            String token = userMap.get(username).getToken();
+            long expiredTime = userMap.get(username).getExpiredTimeMills();
+            if (!needRefresh(expiredTime)) {
+                return token;
+            }
+        }
+        String token = jwtTokenManager.createToken(username);
+        NacosUser user = jwtTokenManager.parseToken(token);
+        long expiredTime = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(getTokenValidityInSeconds());
+        Authentication authentication = jwtTokenManager.getAuthentication(token);
```
Comment (YunWZ) [inline plugin-default-impl/src/main/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManager.java:95]: I don't think we need to save `Authentication` always.
`Authentication` will only be used if the authentication plugin is enabled and the plugin name is not `nacos` or `ldap`

File: plugin-default-impl/src/main/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManager.java (hunk 2 above)
Comment (MajorHe1) [reply plugin-default-impl/src/main/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManager.java:49]: yes, it aims to reduce token quantity

File: plugin-default-impl/src/test/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManagerTest.java (hunk 1 above)
Comment (MajorHe1) [reply plugin-default-impl/src/test/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManagerTest.java:56]: 不知道咋搞，反正抄的 UserController 的搞法 ：）

File: plugin-default-impl/src/main/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManager.java (hunk 3 above)
Comment (MajorHe1) [reply plugin-default-impl/src/main/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManager.java:95]: I really didn't understand the meaning and purpose of Authentication, I just saw that NacosAuthManager called this method.
If it is determined to be useless, can you submit another PR after this PR merge?

**Comment by YunWZ** [comment]
LGTM

File: plugin-default-impl/src/test/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManagerTest.java (hunk 1 above)
Comment (MajorHe1) [reply plugin-default-impl/src/test/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManagerTest.java:56]: 你可以后续搞个PR改成构造器注入的形式吗？让我也看一下怎么改

Comment (CherishCai) [reply plugin-default-impl/src/test/java/com/alibaba/nacos/plugin/auth/impl/token/impl/CachedJwtTokenManagerTest.java:56]: 形如👇，那么 new 对象的时候，也可以传入自定义的了
image

**Review by CherishCai — APPROVED** [review]

**Pull request merged**